File: 1823. Find the Winner of the Circular Game.go
func findTheWinner(n int, k int) int {
    out := make([]bool, n)
    current := 0
    remaining := n
    for {
        if remaining == 1 { break }
        count := 1
        for count < k {
            current++
            current %= n
            if out[current] { continue }
            count++
        }
        out[current] = true
        remaining--
        for {
            current++
            current %= n
            if !out[current] { break }
        }
    }
    return current+1
}

File: 1823. Find the Winner of the Circular Game.java
class Solution {
    public int findTheWinner(int n, int k) {
        Node head = getLinkedList(n);
        Node pointer = head;
        while (pointer.next != pointer) {
            int count = k-1;
            while (count-- > 0) {
                pointer = pointer.next;
            }
            pointer.previous.next = pointer.next;
            pointer.next.previous = pointer.previous;
            pointer = pointer.next;
        }
        return pointer.index;
    }
    private Node getLinkedList(int n) {
        Node head = new Node(1);
        Node pointer = head;
        for (int i = 2; i <= n; i++) {
            Node friend = new Node(i);
            pointer.next = friend;
            friend.previous = pointer;
            pointer = pointer.next;
        }
        pointer.next = head;
        head.previous = pointer;
        return head;
    }
}

class Node {
    Node previous;
    Node next;
    int index;
    Node(int index) {
        this.index = index;
    }
}
